File: week03/practice/http-1/main.go
package main

import (
	"encoding/json"
	"fmt"
	"io"
	"net/http"
)

const todosURL = "https://jsonplaceholder.typicode.com/todos"

type Todo struct {
	Title     string `json:"title"`
	UserID    int    `json:"userId"`
	ID        int    `json:"id"`
	Completed bool   `json:"completed"`
}

func main() {
	client := &http.Client{}

	req, err := http.NewRequest(http.MethodGet, todosURL, nil)
	if err != nil {
		fmt.Printf("build request failed: %v\n", err)
		return
	}

	resp, err := client.Do(req)
	if err != nil {
		fmt.Printf("request failed: %v\n", err)
		return
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		body, _ := io.ReadAll(resp.Body)
		fmt.Printf("request failed: %d %s\n", resp.StatusCode, string(body))
		return
	}

	var todos []Todo
	if err = json.NewDecoder(resp.Body).Decode(&todos); err != nil {
		fmt.Printf("parse todos failed: %v\n", err)
		return
	}

	for _, todo := range todos {
		fmt.Printf(
			"title: %s, userId: %d, id: %d, completed: %t\n",
			todo.Title,
			todo.UserID,
			todo.ID,
			todo.Completed,
		)
	}
}

File: week03/practice/http-1/main.js
const TODOS_URL = "https://jsonplaceholder.typicode.com/todos";

async function run() {
  const response = await fetch(TODOS_URL);
  if (!response.ok) {
    throw new Error(`request failed: ${response.status} ${response.statusText}`);
  }

  const todos = await response.json();
  for (const todo of todos) {
    console.log(
      `title: ${todo.title}, userId: ${todo.userId}, id: ${todo.id}, completed: ${todo.completed}`
    );
  }
}

run().catch((error) => {
  console.error("fetch todos failed:", error.message);
  process.exit(1);
});
